glommio/src/io/immutable_file.rs

```diff
@@ -0,0 +1,395 @@
+    P: AsRef<Path>,
+{
+    /// Creates a new [`ImmutableFileBuilder`].
+    ///
+    /// If this is a new file, use [`build_sink`] to build the
+    /// [`ImmutableFileSink`] stage that will allow writes until [`seal`].
+    /// In that case, the filename must not exist. However, no error happens
+    /// until [`build_sink`] is attempted.
+    ///
+    /// If this is an existing file that will be exclusively read, use
+    /// [`build_existing`] to create an [`ImmutableFile`] directly, skipping
+    /// the [`ImmutableFileSink`] stage
+    ///
+    /// [`ImmutableFile`]: ImmutableFile
+    /// [`ImmutableFileSink`]: ImmutableFileSink
+    /// [`seal`]: ImmutableFileSink::seal
+    /// [`build_sink`]: ImmutableFileBuilder::build_sink
+    /// [`build_existing`]: ImmutableFileBuilder::build_existing
+    pub fn new(fname: P) -> Self {
+        Self {
+            path: fname,
+            buffer_size: 128 << 10,
+            concurrency: 10,
+            max_merged_buffer_size: 1 << 20,
+            max_read_amp: Some(128 << 10),
+            flush_disabled: false,
+        }
+    }
+
+    /// Define the number of buffers that will be used by the
+    /// [`ImmutableFile`] during sequential access.
+    ///
+    /// This number is upheld for both the sink and read phases of the
+    /// [`ImmutableFile`]. Higher numbers mean more concurrency but also
+    /// more memory usage.
+    ///
+    /// [`ImmutableFile`]: ImmutableFile
+    pub fn with_sequential_concurrency(mut self, concurrency: usize) -> Self {
+        self.concurrency = concurrency;
+        self
+    }
+
+    /// Does not issue a sync operation when sealing the file. This is dangerous
+    /// and in most cases may lead to data loss.
+    ///
+    /// Please note that even for `O_DIRECT` files, data may still be present in
+    /// your device's internal cache until a sync happens.
+    ///
+    /// This option is ignored if you are creating an [`ImmutableFile`] directly
+    /// and skipping the [`ImmutableFileSink`] step
+    ///
+    /// [`ImmutableFile`]: ImmutableFile
+    /// [`ImmutableFileSink`]: ImmutableFileSink
+    pub fn with_sync_on_close_disabled(mut self, flush_disabled: bool) -> Self {
+        self.flush_disabled = flush_disabled;
+        self
+    }
+
+    /// Define the buffer size that will be used by the sequential operations on
+    /// this [`ImmutableFile`]
+    ///
+    /// [`ImmutableFile`]: ImmutableFile
+    pub fn with_buffer_size(mut self, buffer_size: usize) -> Self {
+        self.buffer_size = buffer_size;
+        self
+    }
+
+    /// Builds an [`ImmutableFileSink`] with the properties defined by this
+    /// builder.
+    ///
+    /// The resulting sink can be written to until [`seal`] is called.
+    ///
+    /// [`ImmutableFile`]: ImmutableFileSink
+    /// [`new`]: ImmutableFileBuilder::new
+    /// [`seal`]: ImmutableFileSink::seal
+    pub async fn build_sink(self) -> Result<ImmutableFileSink> {
+        println!("siink builder");
```

Comment:
nit

Reply:
Or "niit", I should say :)

Reply:
yiikes! tnx

Reply:
this is a big disadvantage of the tests capturing output. But overall still best this way.

Thanks for catching